shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -0,0 +1,344 @@
+package scalajson
+
+package object ast {
+
+  /**
+    * A regex that will match any valid JSON number for unlimited
+    * precision
+    */
+  @inline protected[ast] val jNumberRegex: String =
+    """-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?"""
+
+  /**
+    * Finds the hashcode for a numeric JSON string.
+    *
+    * @author Matthew de Detrich
+    * @param value
+    * @return
+    */
+  private[ast] def numericStringHashcode(value: String): Int = {
+    var result = 31
+    val length = value.length
+    var i = 0
+
+    if (value(0) == '-') {
+      // Found a negative, increment by one
+      result = result * 31 + '-': Int
+      i = 1
+    }
+
+    var char = value(i)
+
+    // From now on, we can just traverse all the chars
+
+    var negativeFlag = false
+
+    while (i < length) {
+      char = value(i)
+      // if char is e, lowercase it
+      if ((char | 0x20) == 'e') {
+
+        if (value(i + 1) == '-') {
+          // Found a negative, increment by one
+          i += 1
+          char = value(i)
+          negativeFlag = true
+        } else if (value(i + 1) == '+') {
+          // Found a positive, ignore
+          i += 1
+          char = value(i)
+        }
+
+        i += 1
+        char = value(i)
+
+        // Need to skip all leading zeroes, possible with e
+        while (char == '0' && i < length) {
+          i += 1
+          if (i != length)
+            char = value(i) // Fencepost, possible that this can be last character
+        }
+
+        if (i < length) {
+          if (negativeFlag) {
+            result = result * 31 + '-': Int
+          }
+
+          result = 31 * result + 'e': Int
+        }
+      } else if (char == '.') {
+
+        i += 1
+        char = value(i)
+
+        while (char == '0' && i < length) {
+          i += 1
+          if (i != length)
+            char = value(i) // Fencepost, possible that this can be last character
+        }
+
+        if (i < length) {
+          result = 31 * result + '.': Int // The decimal is not finishing with a 0
+        }
+      } else {
+        result = 31 * result + char: Int
+      }
+
+      i += 1
+    }
+    result
+  }
+
+  /** Tests two non-empty strings for equality, assuming both are decimal representations of numbers.
+    *
+    * Note: if the two strings are NOT decimal representations of numbers, the results of this method are undefined.
+    * (It is likely but not guaranteed that the method will return `false` even if the two strings are identical.)
+    *
+    * Many thanks for @Ichoran (https://github.com/Ichoran) for the implementation
+
+    * @author Rex Kerr
+    * @see https://github.com/Ichoran/kse/blob/master/src/main/scala/jsonal/Jast.scala#L683-L917
+    */
+  private[ast] def numericStringEquals(a: String, b: String): Boolean = {
```

Comment:
People ask for (and will ask for) many things.
The hard part of this effort is to know which requests to fulfill and which to deny.
Fulfilling all requests is a sure path to disaster.

IMHO there are two choices:
- Make `JNumber` represent a parsed number, i.e. lift the level of abstraction from the character level to some higher number-representing data type.
- Make `JNumber` a *character representation* of a number, i.e. without actually raising the level of abstraction. Rather: Merely group a bunch of characters and mark them as representing a number.

From reading the source one gets the impression that scalajson is doing the latter and not the former. The former is possible but a bit more involved. Circe goes down this route and is a good example of how it can be done. Unfortunately this requires the introduction of another number type (`BiggerDecimal` in circe's case) since there is no readily available one that can properly represent all JSON numbers.

My impression was that scalajson chose to go down the "stay low-level" path exactly to circumvent these kinds of issues. The consequence would have to be that `JNumber("0") != JNumber("-0")`, which is not only expected from the type setup, it is also consistent with the semantic approach.

I really see no value in going for a strange semantic mix. How would that be useful?
When would I want to rely on character-level equality when I really want some kind of higher-level equality?

The question of whether it's even feasible to properly do equality on arbitrary JSON numbers is a second issue. Is there a proper testsuite for the currently implemented "magic" equality implementation?